packages/web3-eth/src/utils/watch_transaction_for_confirmations.ts

```diff
@@ -61,4 +56,94 @@ export function watchTransactionForConfirmations<
 			}
 		})() as unknown;
 	}, web3Context.transactionReceiptPollingInterval ?? web3Context.transactionPollingInterval);
+};
+
+const watchBySubscription = ({
+	web3Context,
+	transactionReceipt,
+	transactionPromiEvent,
+	returnFormat,
+}: WaitProps) => {
+	setImmediate(() => {
+		web3Context.subscriptionManager
+			?.subscribe('newHeads')
+			.then((subscription: NewHeadsSubscription) => {
+				subscription.on('data', async (data: BlockOutput) => {
+					if (!data?.number) {
+						return;
+					}
+					const confirmationNumber =
+						BigInt(data.number) - BigInt(transactionReceipt.blockNumber) + BigInt(1);
+					transactionPromiEvent.emit('confirmation', {
+						confirmationNumber: format(
+							{ eth: 'uint' },
+							confirmationNumber,
+							returnFormat,
+						),
+						receipt: transactionReceipt,
+						latestBlockHash: format(
+							{ eth: 'bytes32' },
+							data.parentHash as HexString32Bytes,
+							returnFormat,
+						),
+					});
+					if (confirmationNumber >= web3Context.transactionConfirmationBlocks) {
+						await subscription.unsubscribe();
+					}
+				});
+				subscription.on('error', async () => {
+					await subscription.unsubscribe();
+					watchByPolling({
+						web3Context,
+						transactionReceipt,
+						transactionPromiEvent,
+						returnFormat,
+					});
+				});
+			})
+			.catch((err: Error) => {
+				throw new Error(
```

Comment:
Good to use web3 errors [like](https://github.com/ChainSafe/web3.js/blob/4.x/packages/web3-common/src/errors.ts).

Reply:
and also add `transactionReceipt` in error details/message for which subscribe failed.